pkg/util/podstore/podstore.go

```diff
@@ -0,0 +1,194 @@
+	DeletionTimestamp *time.Time
+}
+
+// NewPodStoreWithClock creates a Pod Store with a custom clock,
+// which is useful when writing robust unit tests.
+func NewPodStoreWithClock(podInformer cache.SharedIndexInformer, clock clock.WithTicker) *PodStore {
+	s := &PodStore{
+		curPod:       podInformer.GetIndexer(),
+		prevPod:      cache.NewIndexer(podKeyFunc, cache.Indexers{PodIPIndex: podIPIndexFunc}),
+		podsToDelete: workqueue.NewDelayingQueueWithCustomClock(clock, DeleteQueueName),
+		clock:        clock,
+		timestampMap: map[string]*LocalClock{},
+	}
+	podInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    s.onPodCreate,
+			DeleteFunc: s.onPodDelete,
+		})
+	s.curPod.AddIndexers(cache.Indexers{PodIPIndex: podIPIndexFunc})
+	return s
+}
+
+func NewPodStore(podInformer cache.SharedIndexInformer) *PodStore {
+	return NewPodStoreWithClock(podInformer, clock.RealClock{})
+}
+
+func (s *PodStore) onPodCreate(obj interface{}) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		klog.ErrorS(nil, "Error while processing Create Event")
+		return
+	}
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	timeNow := s.clock.Now()
+	s.timestampMap[string(pod.UID)] = &LocalClock{CreationTimestamp: timeNow}
+	klog.V(4).InfoS("Process Pod Create Event", "Pod", klog.KObj(pod))
+}
+
+func (s *PodStore) onPodDelete(obj interface{}) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		var err error
+		pod, err = s.checkDeletedPod(obj)
+		if err != nil {
+			klog.ErrorS(err, "Got error while processing Delete Event")
+			return
+		}
+	}
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	timeNow := s.clock.Now()
+	s.timestampMap[string(pod.UID)].DeletionTimestamp = &timeNow
+	klog.V(4).InfoS("Process Pod Delete Event", "Pod", klog.KObj(pod))
+	s.prevPod.Add(pod)
+	s.podsToDelete.AddAfter(pod, DelayTime)
+}
+
+func (s *PodStore) checkDeletedPod(obj interface{}) (*corev1.Pod, error) {
+	deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		return nil, fmt.Errorf("received unexpected object: %v", obj)
+	}
+	pod, ok := deletedState.Obj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("DeletedFinalStateUnknown object is not of type Pod: %v", deletedState.Obj)
+	}
+	return pod, nil
+}
+
+func (s *PodStore) GetPodByIPAndTime(ip string, startTime time.Time) (*corev1.Pod, bool) {
+	// Query current Pod first
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	curPods, _ := s.curPod.ByIndex(PodIPIndex, ip)
+	if len(curPods) != 0 {
+		pod := curPods[0].(*corev1.Pod)
+		if s.timestampMap[string(pod.UID)].CreationTimestamp.Before(startTime) && s.timestampMap[string(pod.UID)].DeletionTimestamp == nil {
+			klog.V(4).InfoS("Get Pod info from current store", "Pod", klog.KObj(pod))
+			return pod, true
+		}
+	}
+	// Might not be able to find previous Pod in prePods if prevPod.ByIndex is called immediately after
+	// Delete Event happened. The clientState will delete the Pod in the informer before the DeleteFunc is called.
+	prePods, _ := s.prevPod.ByIndex(PodIPIndex, ip)
```

Comment:
should we use our own indexer (as opposed to `podInformer.GetIndexer()`) to avoid this issue?

Reply:
Yes. I've updated it. Thanks.